src/modules/kvs/kvs.c

```diff
@@ -1223,6 +1229,68 @@ static int heartbeat_root_cb (struct kvsroot *root, void *arg)
     return 0;
 }
 
+static void sync_continuation (flux_future_t *f, void *arg)
+{
+    struct kvs_ctx *ctx = arg;
+    struct kvsroot *root;
+
+    if (flux_rpc_get (f, NULL) < 0)
+        flux_log_error (ctx->h, "%s: flux_rpc_get", __FUNCTION__);
+    flux_future_destroy (ctx->sync_f);
+    ctx->sync_f = NULL;
+
+    /* on rank 0, primary namespace has to be here */
+    root = kvsroot_mgr_lookup_root (ctx->krm, KVS_PRIMARY_NAMESPACE);
+    assert (root);
+
+    /* if there are no commits to be processed, shut down timer, we'll
+     * wake it up next time the primary KVS namespace has work to be
+     * done on it */
+    if (kvstxn_mgr_ready_transaction_count (root->ktm) == 0) {
+        flux_watcher_stop (ctx->sync_w);
+        ctx->sync_timer_running = false;
+    }
```

Comment:
Hmmmm, ok, I see your point, instead of the root check here and the queue-ing checks everywhere else, effectively do some timer math to see if the timer should restarted after every commit.  That might be simpler, lemme give it a try.

As an aside, I realized I do have a corner case, in that the current code guarantees to checkpoint within "sync time" after the first commit (or if the KVS has been non-busy, the first commit in awhile in which the timer is woken up), which is probably not correct.  It should be "sync time" after the KVS started (or sync immediately if there hasn't been a sync in > "sync time").

Also, it occurred to me, to remove the wasteful "strcmp" on every namespace string, I could just add a special flag in the `kvsroot` struct to indicate "i am the primary namespace".  I think that'd be useful.